Tigger/Coordinates.py

```diff
@@ -341,28 +341,110 @@ class FITSWCS(FITSWCSpix):
+                    refsky_pixels.append(wcs_pixel[ipixel])
+            self.refsky = self.wcs.wcs_pix2world([refsky_pixels], 0)[0, :]
+
+            # set centre x/y pixels
+            self.xpix0, self.ypix0 = self.refpix[self.ra_axis], self.refpix[self.dec_axis]
+
+            # set x/y scales
+            pixscale = (header["CDELT1"], header["CDELT2"])
+            self.xscale = pixscale[self.ra_axis] * DEG
+            self.yscale = pixscale[self.dec_axis] * DEG
+
+            has_projection = True
             self._l0 = self.refpix[self.ra_axis]
             self._m0 = self.refpix[self.dec_axis]
 
         def lm(self, ra, dec):
-            l, m = Projection.FITSWCSpix.lm(self, ra, dec)
-            return sin((l - self._l0) * self.xscale), sin((m - self._m0)*self.yscale)
+            if not self.has_projection():
+                return -numpy.sin(ra) / self.xscale, numpy.sin(dec) / self.yscale
+            if numpy.isscalar(ra) and numpy.isscalar(dec):
+                if ra - self.ra0 > math.pi:
+                    ra -= 2 * math.pi
+                if ra - self.ra0 < -math.pi:
+                    ra += 2 * math.pi
+                skyvec = np.array(self.refsky).copy()
+                skyvec[self.ra_axis] = ra / DEG
+                skyvec[self.dec_axis] = dec / DEG
+                pixvec = self.wcs.wcs_world2pix([skyvec], 0)[0]
+                if np.isnan(np.sum(pixvec)):
+                    l, m = -0.0, 0.0
+                else:
+                    l, m = pixvec[self.ra_axis], pixvec[self.dec_axis]
+            else:
+                if numpy.isscalar(ra):
+                    ra = numpy.array(ra)
+                if numpy.isscalar(dec):
+                    dec = numpy.array(dec)
+                n = max(len(ra), len(dec))
+                skymat = numpy.array([self.refsky for _ in range(n)])
+                skymat[:, self.ra_axis] = ra / DEG
+                skymat[:, self.dec_axis] = dec / DEG
+                ra = skymat[:, self.ra_axis]
+                ra[ra - self.ra0 > 180] -= 360
+                ra[ra - self.ra0 < -180] += 360
+                ## when fed in arrays of ra/dec, wcs.wcs2pix will return a nested list of
+                ## [[l1,m1],[l2,m2],,...]. Convert this to an array and extract columns.
+                lm = self.wcs.wcs_world2pix([skymat], 0)
+                l, m = lm[:, self.ra_axis], lm[:, self.dec_axis]
+            l = (l - self._l0) * self.xscale
+            m = (m - self._m0) * self.yscale
+            return l, m
 
         def radec(self, l, m):
-            pixvec = np.array(self.refpix).copy()
-            pixvec[self.ra_axis] = (self.xpix0 - l / self.xscale) + self._l0
-            pixvec[self.dec_axis] = (self.ypix0 + m / self.yscale) + self._m0
-            skyvec = self.wcs.wcs_pix2world([pixvec], 0)[0]
-            ra, dec = skyvec[self.ra_axis], skyvec[self.dec_axis]
+            if not self.has_projection():
+                return numpy.arcsin(-l), numpy.arcsin(m)
```

Comment:
Again I think this is not right:
![image](https://user-images.githubusercontent.com/7934586/117724537-2f59b480-b1e4-11eb-8f10-d0f94faed6ae.png)